### PR TITLE
Use new JAX index update approach for `AdvancedIncSubtensor`

### PR DESCRIPTION
**Pull request by qipengchen**
This PR uses the new JAX index update approach for `jax_funcify_AdvancedIncSubtensor()`.
Like :
[Use new JAX index update approach when available #864](https://github.com/aesara-devs/aesara/pull/864)

### CODE REVIEW

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Looks like you need to set up `pre-commit` locally, run it, and rebase this branch.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
We need some tests that cover these changes before they can be merge.

Also, feel free to squash these commits.

**Comment by qipengchen** [comment]
Hi, do I need to write a test? Link /test_jax.py seems to have a test already. I'm new to this and don't know how to write a test for it. Sorry about that.

**Comment by brandonwillard** [comment]
> Hi, do I need to write a test? **link/test_jax.py** seems to have a test already. I'm new to this and don't know how to write a test for it. Sorry about that.

Yes, there's a test, but it looks like we'll need to augment the test setup so that we have the correct JAX version (or multiple versions) and get coverage for the new lines of code.

Regardless, there's no need to close this PR; you've gotten things started and that's helpful.  If you can't add tests right now, this can stay open until you can.  

Likewise, we can always add the tests to this branch when we get a chance&mdash;in which case, you'll need to `git rebase` locally to pull in those changes, if you want to make further changes on top of them.

In the meantime, feel free to ask questions and we'll try to clarify the process and/or requirements.

**Comment by qipengchen** [comment]
> > Hi, do I need to write a test? **link/test_jax.py** seems to have a test already. I'm new to this and don't know how to write a test for it. Sorry about that.
> 
> Yes, there's a test, but it looks like we'll need to augment the test setup so that we have the correct JAX version (or multiple versions) and get coverage for the new lines of code.
> 
> Regardless, there's no need to close this PR; you've gotten things started and that's helpful. If you can't add tests right now, this can stay open until you can.
> 
> Likewise, we can always add the tests to this branch when we get a chance—in which case, you'll need to `git rebase` locally to pull in those changes, if you want to make further changes on top of them.
> 
> In the meantime, feel free to ask questions and we'll try to clarify the process and/or requirements.

Thanks a lot!  
I lacked the basics knowledge of testing, so I wanted to try running test-related programs locally. How can I run tests locally? Are there any packages or extensions I need to know about?

**Comment by brandonwillard** [comment]
> Thanks a lot!
> I lacked the basics knowledge of testing, so I wanted to try running test-related programs locally. How can I run tests locally? Are there any packages or extensions I need to know about?

Try performing the development setup from [here](https://aesara.readthedocs.io/en/latest/dev_start_guide.html#setting-up-the-your-local-development-environment).  After that, you should be able to run the existing tests with `pytest`.

Running _all_ the tests can take a long time, so the standard development workflow usually involves only running the tests that are clearly relevant to the things that were changed in the PR.  In this case, you should only need to run the tests in the `tests.link.test_jax` module.

Once those tests pass, you can push your local changes to the remote branch that corresponds to this PR and let the CI scripts run the rest of the tests.  If there's a failure in CI, then you can find out exactly where (e.g. a seemingly unrelated test) by reading the logs and attempt to reproduce the issue locally.

See the [`pytest`](https://docs.pytest.org/en/7.1.x/) documentation for further information about test running.

**Comment by qipengchen** [comment]
> 

Thank you for your kind and detailed reply, which was very helpful to me.

**Comment by twiecki** [comment]
@qipengchen Any progress on this?

**Comment by qipengchen** [comment]
> @qipengchen Any progress on this?

I still don't understand that **jax_funcify_IncSubtensor** is covered by the test and **jax_funcify_AdvancedIncSubtensor** is not covered by the test.
And **test_jax_IncSubtensor** is marked as a failure,  I don't know how to debug IncSubtensor tests.

**Comment by qipengchen** [comment]
> 

Sorry for the late reply.

In fact, I tried to run each one separately, however, in `test_jax_IncSubtensor` , an error occurred whenever the line `compare_jax_and_py(out_fg, [])` was run. 

Error contents:  https://github.com/qipengchen/aesara/blob/main/tests/link/try_test_jax.ipynb

This error appears to come from `aesara_jax_fn(*test_inputs)`
https://github.com/aesara-devs/aesara/blob/e2e2366850664544030c1fd769d6b6d25b9d9bba/tests/link/test_jax.py#L102-L104

 In addition, the pytest I run locally, only the first `compare_jax_and_py(out_fg, [])`  of  `test_jax_IncSubtensor`  is coverage by test.

**Comment by brandonwillard** [comment]
> In fact, I tried to run each one separately, however, in `test_jax_IncSubtensor` , an error occurred whenever the line `compare_jax_and_py(out_fg, [])` was run.

The error you're seeing looks like a common issue we've observed with JAX's generalizability/scope (e.g. see [this `arange` issue](https://github.com/aesara-devs/aesara/issues/68) and [this earlier issue](https://github.com/aesara-devs/aesara/issues/43)).

**Comment by brandonwillard** [comment]
> In addition, the pytest I run locally, only the first `compare_jax_and_py(out_fg, [])` of `test_jax_IncSubtensor` is coverage by test.

Which version of `jax` and `jaxlib` were you using?

**Comment by qipengchen** [comment]
> > In addition, the pytest I run locally, only the first `compare_jax_and_py(out_fg, [])` of `test_jax_IncSubtensor` is coverage by test.
> 
> Which version of `jax` and `jaxlib` were you using?

My version is `jax=0.3.7`  and `jaxlib=0.3.7`.

**Comment by brandonwillard** [comment]
> My version is `jax=0.3.7` and `jaxlib=0.3.7`.

It looks like [the version of `jaxlib` used in CI](https://github.com/aesara-devs/aesara/runs/6203686980?check_suite_focus=true#step:5:280) is older (i.e. 0.3.2).  I don't know if `jaxlib` makes the difference in this case, but that's an important distinction.

Also, we can't merge this PR with the notebook included.  For things like that, it's best to create [a Gist](https://gist.github.com/) and link to it in a comment.

**Comment by qipengchen** [comment]
> 
> It looks like [the version of `jaxlib` used in CI](https://github.com/aesara-devs/aesara/runs/6203686980?check_suite_focus=true#step:5:280) is older (i.e. 0.3.2). I don't know if `jaxlib` makes the difference in this case, but that's an important distinction.
> 
> Also, we can't merge this PR with the notebook included. For things like that, it's best to create [a Gist](https://gist.github.com/) and link to it in a comment.

Thanks for your help, I tried installing jaxLib 0.3.2 and running the test code.
The odd thing is to be able to run jax-based algorithms in PyMC, but not work properly in the tests.
https://gist.github.com/qipengchen/c0e85e94f4aacf1c9ecc89fe7d17483b

**Comment by ricardoV94** [comment]
@qipengchen I forced pushed some changes to your branch. I separated the AdvancedIncSubtensor tests that seem to be failing due to JAX limitations from those that were working

File: tests/link/test_jax.py
Comment (ricardoV94) [inline tests/link/test_jax.py:765]: This should have an `xfail` mark to be obvious that it's a failure and not desired behavior.

**Pull request merged**
